### PR TITLE
fix(mcp): read the expiry key the writers actually write

### PR DESCRIPTION
**Pull request by riatzukiza**
## What is broken

ChatGPT now reaches the consent page and approves — then fails. The token
exchange answers `Unknown or expired code` for a code Knoxx minted seconds
earlier:

```
GET  /api/mcp/oauth/authorize/confirm?...   302   (code issued)
POST /api/mcp/oauth/token                   400   "Unknown or expired code"
```

## Why

The writers store `:expiresAt`. The readers asked for `:expires-at` — a key no
document has ever carried — so the `(… 0)` default made the comparison
`(> 0 now)`, false for **every** record.

Confirmed against a live code document in production:

```
keys: _id, code, code_data, created_at, expiresAt, system_instance_id
expiresAt: 2026-08-03T01:38:30.580Z   expires_at: undefined   expires-at: undefined
```

Three readers were affected, all on the critical path:

| reader | effect |
|---|---|
| `get-code!` | every exchange fails as "expired" |
| `get-token!` | no access token can ever be presented — every authenticated `/mcp` call 401s |
| `list-tokens-for-membership!` | always empty |

So even had the exchange succeeded, the token would not have verified on the
next request. Both halves of the remaining flow were dead.

## The change

Reads `:expiresAt` through a shared `live?` helper that accepts the BSON date
the driver returns, plus a number or ISO string for hand-written or migrated
documents, and treats an unreadable expiry as expired so the check fails closed.

`@workspace/mcp-oauth` — the reference that works with ChatGPT — uses `expiresAt`
throughout (`parsed.expiresAt <= now`), confirming the writers here were right
and the readers were wrong.

## Why nothing caught it

Same shape as the `get-client!` envelope bug: a writer and a reader that were
only ever exercised together against a live Mongo, so nothing local could see
them disagree. This is the third instance of that family in this module.

The new tests pin each pair with a double — a freshly written code and token
read back, an expired one does not, and a document with no readable expiry does
not. Restoring the `:expires-at` read fails four assertions:

```
FAIL in (a-freshly-written-code-reads-back)   expected: (some? raw)  actual: (not (some? nil))
FAIL in (a-freshly-written-token-reads-back)  expected: (some? raw)  actual: (not (some? nil))
```

705 tests / 2047 assertions, 0 failures. `clj-kondo` clean.

## What remains unverified

This should carry the flow through code exchange to a live token. The first
authenticated `POST /mcp` with that token is then the last unexercised step —
`get-token!` was dead before this, so no bearer token has ever verified in
production.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed OAuth authorization codes and access tokens being incorrectly treated as expired after storage.
  * Improved expiry handling for numeric, date, and ISO-formatted timestamps.
  * Records with missing or invalid expiry data are now safely rejected.
  * Authorization codes are now securely consumed only once during token exchange.
  * Added safeguards against invalid database queries and malformed stored data.

* **Tests**
  * Added coverage for token revocation, expiry handling, single-use authorization codes, credential validation, and client registration flows.

### CODE REVIEW

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs [hunk 1]
```diff
@@ -34,6 +34,34 @@
...
+
+   The writers store :expiresAt as a BSON date, which the driver hands back as a
+   js/Date; a number or an ISO string are accepted too so a hand-written or
+   migrated document still reads. Anything else yields nil."
+  [doc]
+  (let [v (:expiresAt doc)]
+    (cond
+      (number? v)             v
+      (instance? js/Date v)   (.getTime v)
+      (string? v)             (let [t (.parse js/Date v)] (when-not (js/isNaN t) t))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Fixed in 2cb3d60e — same boundary as the `delete-one!` finding on #214, and I should have applied that precedent here without being told.

Added `extern-mongo/instant-ms`, which decodes a stored BSON instant to epoch milliseconds and owns every driver shape:

| stored value | decoded |
|---|---|
| `js/Date` (what Mongo returns) | epoch ms |
| number | passed through |
| ISO string | parsed — migrated/hand-written docs still read |
| invalid date, non-date, nil | `nil` |

The invalid-date case matters beyond tidiness: `.getTime` on an invalid `Date` is `NaN`, and letting that through would have made an unreadable instant compare as `false` for the wrong reason instead of being explicitly unreadable. It now decodes to `nil`, and `live?` fails closed on that deliberately.

`live?` is left deciding only what a decoded number means. One `js/Date` reference remains in it — `(.now js/Date)`, reading the clock rather than the driver's shape. I read that as outside the boundary this rule governs, but say so if you'd rather the current time were injected too.

Decode coverage added for every branch including invalid dates. 706 tests / 2054 assertions, 0 failures; clj-kondo clean.

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs [hunk 2]
```diff
@@ -34,6 +34,34 @@
...
+      :else                   nil)))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Declining the backfill, with evidence from production:

```
knoxx_mcp_codes    total=0   non-Date expiresAt=0   missing=0
knoxx_mcp_tokens   total=0   non-Date expiresAt=0   missing=0
```

Your reasoning about TTL is correct — MongoDB only expires BSON dates, so a numeric or ISO-string `expiresAt` would linger after `live?` rejected it. But the population is empty and cannot grow: `set-code!` and `set-token!` have only ever written `(js/Date. …)`, and nothing else writes these collections. The number and string branches in `instant-ms` are defensive for hand-written or migrated documents that do not exist.

So a bounded backfill would iterate zero documents forever. If a non-Date writer is ever introduced, this becomes real — and the contract now makes that visible, since such a document decodes and then fails `live?` rather than silently misbehaving.

Worth noting *why* both collections are empty: TTL swept the codes, and no token was ever successfully issued — which is the bug this PR fixes.

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs [hunk 3]
```diff
@@ -34,6 +34,34 @@
...
+   asked for :expires-at, which no document has ever carried, so the default of
+   0 made every code and every token read as already expired: the token
+   exchange answered 'Unknown or expired code' for codes it had just minted,
+   and no access token could ever be presented successfully. An unreadable or
+   missing expiry counts as expired, so this fails closed."
+  [doc]
+  (if-let [ms (expiry-ms doc)]
+    (> ms (.now js/Date))
+    false))
+
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Declining this one, with evidence — and I'd already flagged the same line to Codex rather than quietly deciding it.

`(.now js/Date)` appears at **59 call sites across 32 non-extern files** in this backend. Routing this single one through an adapter would not establish a convention, it would create one inconsistency.

More importantly `extern.mongo` is the wrong owner. AGENTS.md asks for adapters "named for the system they own" — the clock is not Mongo's. Putting `now-ms` there would mean any namespace needing the time reaches for the Mongo adapter, which is a worse boundary than the one being fixed.

The Mongo-shaped part of this — decoding the driver's `Date` — did move to `extern.mongo` in 2cb3d60e. What is left in `live?` is reading the clock and comparing two numbers.

If you want the clock behind a boundary I agree it is defensible, but it should be its own adapter (or injected time) applied across all 59 sites as a deliberate change. Happy to open that separately; it does not belong in an OAuth hotfix.

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs (hunk 1 above)
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Fixed in e250ff9b. Valid and the sharpest of the three — `Infinity` is a number, it compares greater than every clock reading, and unguarded it would leave a code or access token live for good.

Every branch of `instant-ms` now goes through a `finite-instant` guard requiring a finite value inside ECMAScript's representable time range (±8.64e15). `Infinity`, `-Infinity`, `NaN` and out-of-range numbers decode to `nil`, and `live?` already fails closed on `nil`.

Regression coverage added for each, plus an end-to-end case: a token document stamped `expiresAt: Infinity` must not verify through `get-token!`.

File: backend/src/cljs/knoxx/backend/extern/mongo.cljs [hunk 4]
```diff
@@ -12,6 +12,41 @@
...
+
+   Every path goes through finite-instant, so anything that cannot name a real
+   moment — an invalid Date, an unparseable string, Infinity, NaN, a value past
+   the representable range — decodes to nil, and the caller decides what an
+   unreadable instant means."
+  [value]
+  (cond
+    (number? value)           (finite-instant value)
+    (instance? js/Date value) (finite-instant (.getTime value))
+    (string? value)           (finite-instant (.parse js/Date value))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/extern/mongo.cljs]: Fixed in b6917ee7. Fair — I moved the *decode* behind the adapter last round but left the admissible range as an inline test, so the boundary still named no contract.

Added `knoxx.backend.law.mongo` with `EpochMillis`: a number within ECMAScript's representable time range. The bounds are where the security property lives — `Infinity` is a number that compares greater than every clock reading, so an expiry carrying it would leave a credential live for good, and the range rejects it along with `NaN`. `instant-ms` now checks every decoded value against that contract rather than an ad hoc range test.

Two placement notes:

- **`law.mongo`, not `law.mcp-oauth`.** The contract belongs to the Mongo boundary, not to OAuth — `extern.mongo` is a general adapter and a future non-OAuth caller should get the same guarantee.
- **An extern adapter calling a law validator** is the established shape here rather than something I invented: the six `extern.openplanner-translation-mongo` namespaces already do it.

708 tests / 2070 assertions, 0 failures, 0 warnings; clj-kondo clean. The contract has direct coverage too, so the admissible shape is asserted independently of the decode path that consumes it.

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs [hunk 5]
```diff
@@ -90,7 +112,7 @@
            result (await (.findOne c #js {"code" (str code)}))]
        (when result
          (let [doc (keywordize result)]
-           (when (> (:expires-at doc 0) (.now js/Date))
+           (when (live? doc)
              (js/JSON.stringify (clj->js (:code_data doc))))))))))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Fixed in e0cb2ae4. Correct, and worth stating plainly: the outcome was two valid access tokens from one single-use authorization code.

Added `extern-mongo/find-one-and-delete!` — removes and returns a document as one operation, unwrapping both the driver v6 shape and the older `{value: doc}` — and `store/consume-code!` on top of it. The route now claims the code **before** any validation, so a code that then fails PKCE or the client match is spent rather than left for another attempt. That is deliberate: it denies repeated guesses at the verifier.

Your framing is right that this branch is what makes it reachable — while `get-code!` always returned nil, no exchange ever got as far as the window.

The test races three exchanges against one code. The double's `findOneAndDelete` resolves a single winner, and it also offers `findOne`/`deleteOne` that are deliberately *not* atomic together, so restoring the read-then-delete sequence reproduces the defect rather than erroring:

```
FAIL in (concurrent-exchanges-yield-exactly-one-token)
expected: (= 1 (count winners))
  actual: (not (= 1 3))
```

Three tokens from one code. With the atomic consume: exactly one.

712 tests / 2078 assertions, 0 failures, 0 warnings; clj-kondo clean.

File: backend/src/cljs/knoxx/backend/extern/mongo.cljs [hunk 6]
```diff
@@ -4,14 +4,61 @@
...
+   return it as CLJS data, or nil when nothing matched.
+
+   The atomicity is the point: exactly one concurrent caller can receive the
+   document, so a caller can use this to make a single-use credential single
+   use. A read followed by a separate delete cannot promise that.
+
+   Driver v6 returns the document itself; earlier versions wrapped it as
+   {value: doc}. Both are unwrapped here so the caller never has to know."
+  [collection-handle query]
+  (let [result (await (.findOneAndDelete collection-handle (clj->js query)))]
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/extern/mongo.cljs]: Fixed in ef6d8f4a. Fair — I added the boundary in the previous commit and did not give it the contract treatment the instant decode had already received two commits earlier.

`law.mongo` gains two contracts:

- **`FieldEqualityQuery`** — a non-empty map of keyword to scalar. Non-emptiness is the load-bearing part: an empty query matches every document, so on a delete it is the difference between claiming one authorization code and removing all of them. Scalars only, because these call sites are equality matching; an operator map is a different contract and should not arrive here unnoticed.
- **`DecodedDocument`** — a keyword-keyed map, or nothing.

I applied the query guard to `delete-one!` as well, not just the path you flagged. Contracting one of two sibling delete paths seemed the more confusing outcome, and the same empty-query hazard applies to both.

Your point about *where* the output check matters is the sharpest part of the finding, and I have written it into the code: on the code path the record is already deleted by the time the result is decoded, so a caller that misread it could not recover the code it destroyed. An undecodable document now raises rather than being returned.

Tests assert the contracts directly and that neither delete path issues anything to the collection when the query is empty.

714 tests / 2086 assertions, 0 failures, 0 warnings; clj-kondo clean.

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs [hunk 7]
```diff
@@ -34,6 +34,28 @@
...
+   shape; this only decides what the decoded value means. An unreadable or
+   missing expiry counts as expired, so the check fails closed."
+  [doc]
+  ;; instant-ms yields a finite number or nil, but that guarantee does not
+  ;; cross the namespace boundary for the compiler's inference, so both sides
+  ;; of the comparison are hinted rather than left to warn.
+  (if-let [ms (extern-mongo/instant-ms (:expiresAt doc))]
+    (let [^number expiry ms
+          ^number now    (.now js/Date)]
+      (> expiry now))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Fixed in c0056389, and the shape you prescribed — capture the clock in the effectful layer, pass it with the decoded expiry to a named pure guard — is exactly what I implemented.

`law.mcp-oauth/credential-live?` takes the expiry and the clock reading as arguments. The store reads the time; the guard decides what it means. Three consequences worth naming:

- One rule for every reader. `get-code!`, `consume-code!`, `get-token!` and the token listing now cannot drift on what counts as admissible.
- The rule is testable without a clock, which is why the new tests can assert the boundary case (expiring *exactly* now is not live) directly.
- It fails closed on **both** sides. An unnameable expiry is not live, and neither is one paired with an unreadable clock reading — so `Infinity` cannot outlast every reading and a `nil` cannot read as far future.

This also settles the earlier thread about `(.now js/Date)` sitting in the store, and I think lands better than moving it would have: the clock read stays in the effectful layer where it belongs, while the *decision* it feeds no longer does.

715 tests / 2094 assertions, 0 failures, 0 warnings; clj-kondo clean.

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs [hunk 8]
```diff
@@ -82,17 +99,41 @@
...
+   The claim happens before the code's own validity is judged, so a code that
+   turns out to be expired is still consumed. That is deliberate: an expired
+   code is spent either way, and leaving it readable would invite retries."
+  ([code] (consume-code! (mongo-client/get-db) code))
+  ([db code]
+   (when (and db code)
+     (let [doc (await (extern-mongo/find-one-and-delete!
+                       (codes-coll db) {:code (str code)}))]
+       (when (and doc (live? doc))
+         (js/JSON.stringify (clj->js (:code_data doc))))))))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Fixed in 1aaeacac. Right — I had `extern.mongo` decode the document properly and then immediately undid its work: `clj->js` + `JSON.stringify` in the store, `JSON.parse` in the route.

`consume-code!` now returns the `:code_data` map directly. The exchange reads it with keyword access and `persist-access-token!` takes a CLJS map instead of `aget`-ing a parsed JS object. Zero JSON or interop expressions remain on the code path.

Also removed **`get-code!`** — introducing `consume-code!` took its only caller, so it was dead. Worth removing rather than leaving: a reader that does not consume has no business on this path, which is precisely the defect you found two rounds ago.

One thing I have deliberately *not* changed, so it is visible rather than hidden: **`set-token!` and `get-token!` still speak JSON strings.** Same smell, one collection over. I left it because converting them changes what `load-token-record!` and the authenticated `POST /mcp` path receive, and that is the one step in this flow that has never been exercised against a real token — `get-token!` returned nil unconditionally until this branch. I would rather not refactor the untested path in the same commit that makes it reachable.

Say the word and I will do it here; otherwise it is a clean follow-up once a token has actually verified once in production.

715 tests / 2094 assertions, 0 failures, 0 warnings; clj-kondo clean.

File: backend/src/cljs/knoxx/backend/infra/routes/mcp.cljs [hunk 9]
```diff
@@ -562,17 +566,23 @@
...
-              expected (str (or (aget record "codeChallenge") ""))
+      ;; Claimed, not read: an authorization code is single use, and the delete
+      ;; has to happen in the same operation as the read. Reading first and
+      ;; deleting after issuing left a window in which two concurrent exchanges
+      ;; both saw a live code and both minted a token from it.
+      ;;
+      ;; The claim precedes every check below, so a code that fails PKCE or the
+      ;; client match is spent rather than left for another attempt. That is the
+      ;; intent: it denies an attacker repeated guesses at the verifier.
+      (let [record (await (mongo-mcp/consume-code! code))]
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/routes/mcp.cljs]: Fixed in b9fc6757. You are right and I want to be explicit that this **overturns the ordering I chose two commits ago and defended in this review** — my reasoning there was weak.

I argued that consuming before validating denied an attacker repeated guesses at the verifier. A PKCE verifier is 32 random bytes: 256 bits. It is not guessable within the code's five-minute life whether or not the code survives a failed attempt, so that argument bought nothing at all. Your cost is real and cheap to trigger: anyone who merely observes the code on the front channel can destroy it with any wrong verifier and break the legitimate exchange at will.

The exchange now validates the client, redirect and PKCE bindings against a non-destructive `peek-code!`, and claims only once they hold.

**Single use is unaffected**, because it never depended on the ordering — it depends on the claim being atomic:

- two concurrent exchanges can both pass the bindings
- `findOneAndDelete` picks one winner; the loser gets `invalid_grant` "Authorization code already used"
- the token is minted from the **claimed** record, not the peeked one, so its contents cannot have changed between check and claim

The concurrency test still reports exactly one winner.

One note against myself: `peek-code!` is the reader I deleted last commit as dead, at your prompting. Removing it was premature — what was dead was the *route* reading without consuming, not the capability. Reinstated with a name that states it does not spend.

New tests: peeking twice leaves the code claimable and the legitimate exchange still succeeds, and a peek applies the same liveness rule as a claim.

717 tests / 2099 assertions, 0 failures, 0 warnings; clj-kondo clean.

File: backend/src/cljs/knoxx/backend/infra/routes/mcp.cljs [hunk 10]
```diff
@@ -540,18 +540,39 @@
...
+   code it presents.
+
+   Checked against a non-destructive read so a rejected request never spends the
+   code: anyone who merely observed the code on the front channel could
+   otherwise destroy it with any wrong verifier, and it would buy nothing, since
+   a verifier carries far too much entropy to guess."
+  [crypto record client-id redirect-uri code-verifier]
+  (when (or (not= (:clientId record) client-id)
+            (not= (:redirectUri record) redirect-uri))
+    (throw (http-error 400 "invalid_grant" "Client/redirect mismatch")))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/routes/mcp.cljs]: Fixed in e4f5e725, following the split you set out — parsing and challenge computation stay in the effectful layer, the decision moves to law.

`law.mcp-oauth` gains two pure predicates:

| | |
|---|---|
| `code-bound-to?` | the code was issued to this client and redirect |
| `pkce-verified?` | the computed challenge matches the stored one |

The route keeps what is genuinely effectful — parsing the body and computing the challenge from the verifier, which is crypto — and asks law whether the result admits the exchange. Same shape as `credential-live?`.

Being pure, both could be tested exhaustively without constructing a request, and that surfaced an edge worth naming: **a code record carrying no challenge is never verified.** Previously the route checked `(str/blank? expected)` inline; now it is stated as a rule, and the test says why — without it, a code written without a challenge could be redeemed by omitting the verifier entirely.

718 tests / 2109 assertions, 0 failures, 0 warnings; clj-kondo clean.
